File: python/src/Function/handler.py
import json
import boto3
import pyodbc
import os
from openpyxl import Workbook
from io import BytesIO

# Getting database credentials from env variables
curr_server = os.getenv('DB_SERVER')
curr_database = os.getenv('DB_NAME')
curr_user_id = os.getenv('DB_USER')
curr_pass=os.getenv('DB_PASSWORD')

def handler(event, context):
    try: 
        # Making connection string
        conn_str = f"DRIVER={{ODBC Driver 17 for SQL Server}};SERVER={curr_server};DATABASE={curr_database};UID={curr_user_id};PWD={curr_pass}"

        # Establish a connection to SQL Server
        conn = pyodbc.connect(conn_str)
        cursor = conn.cursor() # Creating a cursor object to interact with the database

        # Executing the stored procedure
        cursor.execute("{CALL dbo.Dummy_sp}")
        rows=cursor.fetchall()

        # Creating a workbook
        wb=Workbook()
        ws=wb.active
        ws.title="Stored procedure Dummy_sp data" # Naming the worksheet

        # Adding the column headers
        columns = [desc[0] for desc in cursor.description]
        ws.append(columns)

        # Processing the rows and replacing None (NULL values) with a placeholder ('NULL')
        for row in rows:
            row_data = [value if value is not None else 'NULL' for value in row]
            ws.append(row_data)

        # Saving the workbook to a BytesIO stream (in-memory file)
        excel_file=BytesIO()
        wb.save(excel_file)
        excel_file.seek(0)
        
        # Uploading the file to S3
        s3 = boto3.client('s3')
        bucket_name = os.getenv('s3_bucket_name')
        file_name='stored_procedure_results.xlsx'

        try:
            s3.put_object(Body=excel_file, Bucket=bucket_name, Key=file_name)
            print(f"File uploaded successfully to {bucket_name}/{file_name}")
        except Exception as s3_error:
            print(f"Error uploading file to S3: {str(s3_error)}")
            return {
                'statusCode': 500,
                'body': f"Error uploading file to S3: {str(s3_error)}"
            }
    except Exception as e:
        print(f"Error: {str(e)}")
        return {
            'statusCode': 500,
            'body': f"Error: {str(e)}"
        }
    finally:
        conn.close()        # Ensuring the database connection is closed regardless of success or failure
        excel_file.close()  # Closing the BytesIO stream to free resources
    
    return {
        "statusCode": 200,
        "body": json.dumps({
            "message": "Connection successful and file uploaded successfully"
        }),
    }



    
   

File: java/src/Function/src/main/java/Handler.java
import com.amazonaws.services.s3.AmazonS3;
import com.amazonaws.services.s3.AmazonS3ClientBuilder;
import com.amazonaws.services.s3.model.PutObjectRequest;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;
import org.apache.poi.ss.usermodel.*;
import com.amazonaws.services.lambda.runtime.Context;
import com.amazonaws.services.lambda.runtime.RequestHandler;

import java.io.*;
import java.sql.*;
import java.util.*;

public class Handler implements RequestHandler<Map<String, Object>, String> {

    @Override
    public String handleRequest(Map<String, Object> event, Context context) {
        // Define environment variables
        String dbServer = System.getenv("DB_SERVER");
        String dbName = System.getenv("DB_NAME");
        String dbUser = System.getenv("DB_USER");
        String dbPassword = System.getenv("DB_PASSWORD");
        String s3BucketName = System.getenv("s3_bucket_name");

        try {
            // Establish database connection
            String connectionUrl = String.format("jdbc:sqlserver://%s:1433;databaseName=%s;user=%s;password=%s",dbServer,dbName, dbUser, dbPassword);
            Connection connection = DriverManager.getConnection(connectionUrl);

            // Execute the stored procedure
            CallableStatement stmt = connection.prepareCall("{CALL dbo.Dummy_sp}");
            ResultSet resultSet = stmt.executeQuery();

            // Create an Excel workbook and sheet
            XSSFWorkbook workbook = new XSSFWorkbook();
            Sheet sheet = workbook.createSheet("Stored procedure Dummy_sp data");

            // Create the header row
            ResultSetMetaData metaData = resultSet.getMetaData();
            Row headerRow = sheet.createRow(0);
            int columnCount = metaData.getColumnCount();
            for (int i = 1; i <= columnCount; i++) {
                headerRow.createCell(i - 1).setCellValue(metaData.getColumnLabel(i));
            }

            // Add data rows
            int rowNum = 1;
            while (resultSet.next()) {
                Row row = sheet.createRow(rowNum++);
                for (int i = 1; i <= columnCount; i++) {
                    Object value = resultSet.getObject(i);
                    if (value == null) {
                        row.createCell(i - 1).setCellValue("NULL");
                    } else {
                        row.createCell(i - 1).setCellValue(value.toString());
                    }
                }
            }

            // Write the Excel file to a ByteArrayOutputStream (in-memory)
            ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
            workbook.write(byteArrayOutputStream);
            byteArrayOutputStream.flush();

            // Upload the file to S3
            AmazonS3 s3Client = AmazonS3ClientBuilder.defaultClient();
            String fileName = "stored_procedure_results.xlsx";
            InputStream inputStream = new ByteArrayInputStream(byteArrayOutputStream.toByteArray());

            PutObjectRequest putObjectRequest = new PutObjectRequest(s3BucketName, fileName, inputStream, null);
            s3Client.putObject(putObjectRequest);

            System.out.println("File uploaded successfully to " + s3BucketName + "/" + fileName);

            // Closing resources
            resultSet.close();
            stmt.close();
            connection.close();
            byteArrayOutputStream.close();
            workbook.close();

            return "File uploaded successfully to " + s3BucketName + "/" + fileName;

        } catch (Exception e) {
            e.printStackTrace();
            return "Error: " + e.getMessage();
        }
    }
}
